File: delegation.go
package demesne

import "sort"

type DelegationCap struct {
	Allowed bool

	Unknown []string

	Excess []string
}

func (v *Vocabulary) CapGrant(held, requested []string) DelegationCap {
	inVocab := make(map[string]bool, len(v.Permissions))
	for _, p := range v.Permissions {
		inVocab[p] = true
	}
	heldSet := make(map[string]bool, len(held))
	for _, p := range held {
		heldSet[p] = true
	}
	var unknown, excess []string
	seenU, seenE := map[string]bool{}, map[string]bool{}
	for _, p := range requested {
		switch {
		case !inVocab[p]:
			if !seenU[p] {
				seenU[p] = true
				unknown = append(unknown, p)
			}
		case !heldSet[p]:

			if !seenE[p] {
				seenE[p] = true
				excess = append(excess, p)
			}
		}
	}
	sort.Strings(unknown)
	sort.Strings(excess)
	return DelegationCap{Allowed: len(unknown) == 0 && len(excess) == 0, Unknown: unknown, Excess: excess}
}

File: ts/packages/runtime/src/delegation.ts


import { goSort } from "./goCompat.js";
import type { Vocabulary, DelegationCap } from "./types.js";

export function capGrant(vocab: Vocabulary, held: readonly string[], requested: readonly string[]): DelegationCap {
  const inVocab = new Set(vocab.permissions);
  const heldSet = new Set(held);
  const unknown: string[] = [];
  const excess: string[] = [];
  const seenU = new Set<string>();
  const seenE = new Set<string>();
  for (const p of requested) {
    if (!inVocab.has(p)) {
      if (!seenU.has(p)) {
        seenU.add(p);
        unknown.push(p);
      }
    } else if (!heldSet.has(p)) {

      if (!seenE.has(p)) {
        seenE.add(p);
        excess.push(p);
      }
    }
  }
  const u = goSort(unknown);
  const e = goSort(excess);
  return { allowed: u.length === 0 && e.length === 0, unknown: u, excess: e };
}
